src/server/routes/reporting-periods.js

```diff
@@ -194,6 +197,37 @@ router.get('/:id/template', requireUser, async (req, res, next) => {
   res.end(Buffer.from(data, 'binary'))
 })
 
+router.post('/:id/revalidate', requireAdminUser, async (req, res, next) => {
+  const periodId = req.params.id
+  const commit = req.query.commit || false
+
+  const user = await getUser(req.signedCookies.userId)
+  const reportingPeriod = await reportingPeriods.get(periodId)
+  if (!reportingPeriod) {
+    res.sendStatus(404)
+    res.end()
+    return
+  }
+
+  const trns = await knex.transaction()
+  try {
+    const updates = await revalidateUploads(reportingPeriod, user, trns)
+    if (commit) {
+      trns.commit()
+    } else {
+      trns.rollback()
+    }
```

Comment:
I'm actually pretty happy with this solution for managing transactions!  Nice work.